### PR TITLE
Make template_id optional in send_email order action with server-side auto-selection

### PR DESCRIPTION
**Pull request by staskus**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Makes `template_id` optional in `POST /orders/{id}/actions/send_email`. When omitted, the server picks the best template based on order status.

A shared `STATUS_TEMPLATE_MAP` constant now drives both template availability and auto-selection, removing the duplicate status-to-template mappings that existed before. POS email classes register their preferred templates through a new filter (`woocommerce_rest_order_actions_email_preferred_template_ids`), following the same self-registering pattern they already use for availability. This keeps the controller POS-agnostic.

Related mobile PRs:
- iOS: https://github.com/woocommerce/woocommerce-ios/pull/16781
- Android: https://github.com/woocommerce/woocommerce-android/pull/15485

### Security considerations

Auto-selection reuses the existing `get_available_email_templates()` validation, so the selected template is always one that was already approved for the given order. No new permissions are introduced. The preference filter only reorders templates that passed availability checks, it cannot inject arbitrary ones. If no preferred template matches, the first available template is used as a fallback. If no templates are available at all, the endpoint returns a `400` error.

### How to test the changes in this Pull Request:

1. Create orders in various statuses (completed, processing, on-hold, failed, refunded, pending, cancelled)
2. Call `POST /wp-json/wc/v4/orders/{id}/actions/send_email` without `template_id` and verify the correct template is auto-selected
3. Call the same endpoint with an explicit `template_id` and verify it still works as before
4. Create a POS order (`created_via: pos-rest-api`) and verify POS templates are preferred during auto-selection
5. Verify error scenarios: wrong-status template, nonexistent template, missing billing email

### Testing that has already taken place:

48 scenarios across 14 order configurations: 27 valid sends, 18 error rejections, 3 email update edge cases. All POS detection paths verified. PHPUnit tests cover auto-selection for every status, POS preference, filter override, and error cases.

<details>
<summary>Detailed test results</summary>

#### Valid sends (27 combinations, all passed)

| Status | Order Types Tested | Templates Verified |
|--------|-------------------|-------------------|
| completed | regular, POS (created_via), POS (card terminal), POS (cash), partial refund | `customer_completed_order`, `customer_pos_completed_order`, `customer_invoice`, `customer_refunded_order` |
| refunded | regular, POS (created_via) | `customer_refunded_order`, `customer_pos_refunded_order`, `customer_pos_completed_order`, `customer_invoice` |
| processing | regular, POS (created_via) | `customer_processing_order`, `customer_pos_completed_order`, `customer_invoice` |
| on-hold | regular | `customer_on_hold_order`, `customer_invoice` |
| failed | regular | `customer_failed_order`, `customer_invoice` |
| pending | regular | `customer_invoice` |
| cancelled | regular | `customer_invoice` |

#### Error scenarios (18 tests, all correctly rejected)

| Scenario | Expected Error |
|----------|---------------|
| Wrong-status template (13 combos) | `woocommerce_rest_invalid_email_template` |
| POS template on non-POS order (2 combos) | `woocommerce_rest_invalid_email_template` |
| Nonexistent template | `rest_invalid_param` |
| No billing email on order | `woocommerce_rest_missing_email` |
| Nonexistent order | `woocommerce_rest_not_found` |

#### Email update edge cases (3 tests)

| Scenario | Result |
|----------|--------|
| Change email without `force_email_update` | Rejected |
| Change email with `force_email_update: true` | Updated and sent |
| Provide email to no-email order | Updated billing email and sent |

</details>

### Milestone

- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.
-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message
Add server-side auto-selection of email templates when template_id is omitted in the send_email order action.

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/emails/class-wc-email-customer-pos-completed-order.php [hunk 1]
```diff
@@ -403,6 +404,24 @@ public function add_to_valid_template_classes( $valid_template_classes, $order )
...
+		 *
+		 * @internal For exclusive usage within this class, backwards compatibility not guaranteed.
+		 *
+		 * @param array    $priority Ordered array of template IDs.
+		 * @param WC_Order $order    The order.
+		 * @return array Modified priority array.
+		 *
+		 * @since 10.7.0
+		 */
+		public function add_to_default_template_priority( $priority, $order ) {
```
Comment (joshheald) [inline plugins/woocommerce/includes/emails/class-wc-email-customer-pos-completed-order.php]:
```suggestion
		public function add_to_default_template_priority( $templates, $order ) {
```

`priority` makes me think it's a hook sequencing priority...

File: plugins/woocommerce/includes/emails/class-wc-email-customer-pos-refunded-order.php [hunk 2]
```diff
@@ -479,6 +480,24 @@ public function add_to_valid_template_classes( $valid_template_classes, $order )
...
+		 *
+		 * @internal For exclusive usage within this class, backwards compatibility not guaranteed.
+		 *
+		 * @param array    $priority Ordered array of template IDs.
+		 * @param WC_Order $order    The order.
+		 * @return array Modified priority array.
+		 *
+		 * @since 10.7.0
+		 */
+		public function add_to_default_template_priority( $priority, $order ) {
```
Comment (joshheald) [inline plugins/woocommerce/includes/emails/class-wc-email-customer-pos-refunded-order.php]: As before. Neither of these are must-do if this is in keeping with the other functions, I'm reviewing it with very little context so yeah... take it all with a pinch of salt!

**Review by joshheald — COMMENTED** [review]
I've had an initial look. It works for the POS and web order use cases that I've tried, but I'll do more extensive testing tomorrow. I don't have a great deal helpful to say about the approach though, it seems reasonable to me. I feel like it could be made easier to follow in places... though I'm not exactly sure how. PHP.

Anyway, good work, I'll check back in tomorrow.

File: plugins/woocommerce/includes/emails/class-wc-email-customer-pos-refunded-order.php (hunk 2 above)
Comment (staskus) [reply plugins/woocommerce/includes/emails/class-wc-email-customer-pos-refunded-order.php]: My first iteration was listing the templates manually, `OrderActionsRestController`, and setting default ones based on status and other parameters. That is super readable but not reusable or extensible. 

Since we already had a method that lists available templates based on status, that was the obvious place to reuse, so `STATUS_TEMPLATE_MAP` was created to reuse in the available and default templates case. 

However, it wouldn't cover optional templates, such as POS or any other ones that plugins introduce. Therefore, allowing to add a default one via `woocommerce_rest_order_actions_email_default_template_priority` seemed like a WordPress-friendly version to me. However, I'm sure one more iteration is possible to make it more robust.

**Comment by joshheald** [comment]
@staskus Am I right in thinking we shouldn't get this error any more?

```
{
  "code": "rest_missing_callback_param",
  "message": "Missing parameter(s): template_id",
  "data": {
    "status": 400,
    "params": [
      "template_id"
    ]
  }
}
```

I got that with this request:

```
curl -X "POST" "https://practically-versatile-dreamland.commerce-garden.com/wp-json/wc/v3/orders/109/actions/send_email" 
```

Using an application password for my user and basic auth. Order 109 was already in the `completed` state, with `pos-rest-api`.

**Comment by staskus** [comment]
@joshheald try again. I was testing another PR, and reverted the version of WooCommerce on this site by accident for a few minutes. Bad timing. I won't touch this site again.

File: plugins/woocommerce/src/Internal/Orders/OrderActionsRestController.php [hunk 3]
```diff
@@ -437,18 +520,32 @@ protected function send_email( WP_REST_Request $request ) {
...
-					// translators: %s is a string ID for an email template.
-					__( '%s is not a valid template for this order.', 'woocommerce' ),
-					esc_html( $template_id )
-				),
-				array( 'status' => 400 )
-			);
+		if ( empty( $template_id ) ) {
+			$template    = $this->select_default_template( $order, $available_templates );
+			$template_id = $template ? $template->id : null;
+
```
[the remark this answers is not in the document]
Comment (staskus) [reply plugins/woocommerce/src/Internal/Orders/OrderActionsRestController.php:534]: This is handled by `PointOfSaleEmailHandler` (`src/Internal/Orders/PointOfSaleEmailHandler.php`). It hooks into `woocommerce_email_enabled_customer_refunded_order` and returns `false` when `PointOfSaleOrderUtil::is_order_paid_at_pos()` is true. So the generic refunded email's `is_enabled()` check blocks it for POS orders. Only the POS-specific email sends.

File: plugins/woocommerce/src/Internal/Orders/OrderActionsRestController.php [hunk 4]
```diff
@@ -368,6 +380,72 @@ private function get_email_template_by_id( string $template_id, ?array $availabl
...
+			$order,
+			array_map( fn( $t ) => $t->id, $available_templates )
+		);
+
+		foreach ( $priority as $candidate_id ) {
+			$template = $this->get_email_template_by_id( $candidate_id, $available_templates );
+			if ( $template ) {
+				return $template;
+			}
+		}
```
[the remark this answers is not in the document]
Comment (staskus) [reply plugins/woocommerce/src/Internal/Orders/OrderActionsRestController.php]: Addressed in 2eeb80cbe6. Added `is_array` fallback and `array_filter( array_unique( ... ), 'is_string' )` sanitization after `apply_filters`, matching the pattern used by the sibling `woocommerce_rest_order_actions_email_valid_template_classes` filter on line 350.

File: plugins/woocommerce/includes/emails/class-wc-email-customer-pos-completed-order.php (hunk 1 above)
Comment (staskus) [reply plugins/woocommerce/includes/emails/class-wc-email-customer-pos-completed-order.php]: [Rename priority to preferred_template_ids for clarity across filter, …](https://github.com/woocommerce/woocommerce/pull/63556/commits/6f2bee635143ededfdb0d388beb0835394b06833)

I renamed it. Let me know if it's clearer.

**Review by joshheald — APPROVED** [review]
I've tested this quite thoroughly, and aside from the few things I shared, found it to work well. Once fixed, it looks good to me.

A review from someone in a core team would still be beneficial of course.

Issues:

- POS orders which are `failed`, `on-hold`, or `cancelled`, use the `customer_pos_completed_order` template incorrectly
- Potentially fine: Orders with metadata `{"key":"_cash_change_amount","value":"0"}` get POS emails, even if they're not. I expect that this is pre-existing, and the app doesn't set it for IPP cash payments, but the meta data doesn't say it's POS specific so I wanted to mention it.

**Comment by staskus** [comment]
## Fix: Restrict POS email templates to matching order statuses

Commit: 7a88b07e597149e8746b0647cd78a96fd21a177f

**Problem:** POS completed and refunded email templates were available and preferred for *any* POS order regardless of status. For example, a failed POS order would auto-select `customer_pos_completed_order` instead of `customer_failed_order`.

**Fix:** Extract `is_applicable_for_order()` in each POS email class to check both POS origin and the correct status/refund condition:
- POS Completed: requires `is_order_paid_at_pos()` AND `status === completed`
- POS Refunded: requires `is_order_paid_at_pos()` AND (`status === refunded` OR has refund records)

### Tested on a site (post-fix) - all 25 scenarios pass

**Table 1: Basic Combinations**

| Status | Created Via | Available Templates | Auto-Select | Correct? |
|---|---|---|---|---|
| completed | regular | customer_completed_order, customer_invoice | Order fulfilled | Good |
| completed | pos-rest-api | customer_completed_order, customer_invoice, customer_pos_completed_order | POS order complete | Good |
| processing | regular | customer_invoice, customer_processing_order | Order confirmation | Good |
| processing | pos-rest-api | customer_invoice, customer_processing_order | Order confirmation | Good |
| on-hold | regular | customer_invoice, customer_on_hold_order | Order on hold | Good |
| on-hold | pos-rest-api | customer_invoice, customer_on_hold_order | Order on hold | Good |
| failed | regular | customer_failed_order, customer_invoice | Order failed | Good |
| failed | pos-rest-api | customer_failed_order, customer_invoice | Order failed | Good |
| pending | regular | customer_invoice | Order details | Good |
| pending | pos-rest-api | customer_invoice | Order details | Good |
| refunded | regular | customer_invoice, customer_refunded_order | Order refunded | Good |
| refunded | pos-rest-api | customer_invoice, customer_pos_refunded_order, customer_refunded_order | POS order refunded | Good |

**Table 2: Refund Combinations**

| Status | Created Via | Refund | Available Templates | Auto-Select | Correct? |
|---|---|---|---|---|---|
| refunded | regular | full | customer_invoice, customer_refunded_order | Order refunded | Good |
| completed | regular | partial | customer_completed_order, customer_invoice, customer_refunded_order | Order fulfilled | Good |
| refunded | pos-rest-api | full | customer_invoice, customer_pos_refunded_order, customer_refunded_order | POS order refunded | Good |
| completed | pos-rest-api | partial | customer_completed_order, customer_invoice, customer_pos_completed_order, customer_pos_refunded_order, customer_refunded_order | POS order complete | Good |

**Table 3: Status Change Scenarios**

| # | Scenario | Available Templates | Auto-Select | Correct? |
|---|---|---|---|---|
| 1 | POS completed->failed | customer_failed_order, customer_invoice | Order failed | Good |
| 2 | POS completed->on-hold | customer_invoice, customer_on_hold_order | Order on hold | Good |
| 3 | POS failed->completed | customer_completed_order, customer_invoice, customer_pos_completed_order | POS order complete | Good |
| 4 | POS completed+partial refund | customer_completed_order, customer_invoice, customer_pos_completed_order, customer_pos_refunded_order, customer_refunded_order | POS order complete | Good |
| 5 | POS completed->full refund->refunded | customer_invoice, customer_pos_refunded_order, customer_refunded_order | POS order refunded | Good |
| 6 | POS on-hold->processing->completed | customer_completed_order, customer_invoice, customer_pos_completed_order | POS order complete | Good |
| 7 | Regular completed->failed | customer_failed_order, customer_invoice | Order failed | Good |
| 8 | POS pending->completed->partial refund | customer_completed_order, customer_invoice, customer_pos_completed_order, customer_pos_refunded_order, customer_refunded_order | POS order complete | Good |
| 9 | POS refunded (no refund records) | customer_invoice, customer_pos_refunded_order, customer_refunded_order | POS order refunded | Good |

**Comment by staskus** [comment]
@Konamiman would you have the capacity to check one more email-related PR this week?

We have a previously built `actions/send_email` endpoint that currently requires passing`template_id`. This PR makes it optional by auto-selecting a template based on order status and available templates. It also adds a filter for extensions (like POS) to customize the default selection, using a mechanism already created for defining `valid_template_classes`.

We tested and reviewed on our side, but we want to make sure there are no fundamental flaws. Let me know, thank you.

**Review by Konamiman — APPROVED** [review]

**Comment by Konamiman** [comment]
Hi @staskus, sorry I was in a team meetup this week. The changes look good to me.
